### PR TITLE
Fix for ZD 11808.

### PR DESCRIPTION
**Pull request by haydenroche5**

### CODE REVIEW

**Comment by haydenroche5** [comment]
I'm not sure this is the right fix, actually. When executing the reproducer, just after the call to `GetRecordHeader` at internal.c:15377, we can see the following in gdb:
```
(gdb) where
#0  ProcessReply (ssl=0x621000002900) at src/internal.c:15381
#1  0x000000000057fdb8 in wolfSSL_connect (ssl=0x621000002900) at src/ssl.c:12538
#2  0x000000000054f25c in LLVMFuzzerTestOneInput (data=0x625000002900 "\026\200\377\377\377\200", size=9003) at client.c:70
#3  0x0000000000457d12 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) ()
#4  0x0000000000443483 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) ()
#5  0x0000000000448f37 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) ()
#6  0x0000000000471bf3 in main ()
(gdb) p ssl->buffers.inputBuffer
$9 = {staticBuffer = "\000\000\000\000", buffer = 0x61b000001c80 "\024\376", <incomplete sequence \375>, length = 1511, idx = 1511, bufferSize = 1511, dynamicFlag = 1 '\001', 
  offset = 0 '\000'}
```
At this point, using `ssl->buffers.inputBuffer.idx` as an index is unsafe, because it's out of range (> 1510). It seems safer, then, that we move the check right here, after the call to `GetRecordHeader`. Thoughts?

**Comment by julek-wolfssl** [comment]
The issue with your proposal is that it will too pre-emptively return an error. Notice that `ssl->curSize = 0` after this call so running out of buffer isn't an error in this case. Right after this we have:
```
#ifdef WOLFSSL_DTLS
                /* read ahead may already have */
                used = ssl->buffers.inputBuffer.length -
                       ssl->buffers.inputBuffer.idx;
                if (used < ssl->curSize)
                    if ((ret = GetInputData(ssl, ssl->curSize)) < 0)
                        return ret;
#endif
```
This attempts to read more if we don't have enough buffer to cover the rest of the record. This won't get called since `ssl->curSize = 0` so so far so good. I went through the remaining paths and they will all error out if they exhaust the buffer. The ChangeCipherSpec path is the only one that doesn't first check if it has enough buffer to check for the 1 value.

Now its interesting to note that the RFC states that (https://tools.ietf.org/html/rfc5246#section-6.2.1):
>    Implementations MUST NOT send zero-length fragments of Handshake,
   Alert, or ChangeCipherSpec content types.  Zero-length fragments of
   Application data MAY be sent as they are potentially useful as a
   traffic analysis countermeasure.

So I would support a solution that would do the following:
- return a `LENGTH_ERROR` when `ssl->curSize = 0` after the `GetRecordHeader` call
- in the ChangeCipherSpec check that:
  - `ssl->curSize >= 1`
  - `ssl->buffers.inputBuffer.idx >= buffers.inputBuffer.length` 

Also in the DTLS case, the read ahead `GetInputData` (mentioned above) should be called not with `ssl->curSize` as the size parameter but `ssl->curSize - used`. This should also be fixed as we are forcing the network layer to read more than we actually need.

**Comment by haydenroche5** [comment]
I agree that my proposal will too pre-emptively return an error. Thanks for reasoning through that. Your proposed solution makes sense to me. I'm curious what @ejohnstown thinks.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
As stated in comments.

File: src/internal.c [hunk 1]
```diff
@@ -15377,6 +15377,12 @@ int ProcessReply(WOLFSSL* ssl)
             ret = GetRecordHeader(ssl, ssl->buffers.inputBuffer.buffer,
                                        &ssl->buffers.inputBuffer.idx,
                                        &ssl->curRL, &ssl->curSize);
+
+            if (ssl->curSize == 0 && ssl->curRL.type != application_data) {
```
Comment (haydenroche5) [inline src/internal.c:15381]: I wonder if we should put this in `GetRecordHeader` instead of here after the call site.

Comment (julek-wolfssl) [reply src/internal.c:15381]: Could be placed at the end with the other `LENGTH_ERROR` checks.

**Review by ejohnstown — APPROVED** [review]
I agree with the changes.

**Review by julek-wolfssl — CHANGES_REQUESTED** [review]
Could we also address this issue in this pr?
>Also in the DTLS case, the read ahead GetInputData (mentioned above) should be called not with ssl->curSize as the size parameter but ssl->curSize - used. This should also be fixed as we are forcing the network layer to read more than we actually need.

**Comment by haydenroche5** [comment]
> Could we also address this issue in this pr?
> 
> > Also in the DTLS case, the read ahead GetInputData (mentioned above) should be called not with ssl->curSize as the size parameter but ssl->curSize - used. This should also be fixed as we are forcing the network layer to read more than we actually need.

Sure. Do you want to push that change here on top of mine?

**Comment by julek-wolfssl** [comment]
I think that you can change the call in the DTLS GetInputData.

**Review by julek-wolfssl — APPROVED** [review]

**Review by ejohnstown — APPROVED** [review]

**Pull request merged**